Multiple line description support

darcyparker opened the pull request:
Pull request to fix https://github.com/bcherny/json-schema-to-typescript/issues/24

This replaces my previous pull request that concatenated multiple line descriptions.

bcherny commented on test/cases/with-description-newlines.ts:
```diff
@@ -0,0 +1,30 @@
+  firstName: string; // first name single line description
+  lastName: string;
+  // Age description with
+  // multiple lines
```
Possible to use a multiline comment instead?

darcyparker replied:
Done in https://github.com/bcherny/json-schema-to-typescript/pull/26/commits/adace03e6f19311050885f91185ab39ad5caab45

darcyparker commented on dist/index.js:
```diff
@@ -233,7 +236,10 @@ var TsType;
-                        _.type.description.split(newLineRegex).map(function (line) { return indentString + '// ' + line; }).join('\n') + '\n' + decl :
+                        _.type.description
+                            .split(newLineRegex)
+                            .map(function (line, lineNum) { return indentString + (lineNum > 0 ? multiLineCommentIndent : multiLineCommentStart) + line; })
```
As you know, `index.js` is generated by `gulp dist`. Curiously, it is declaring `function (line, lineNum)` rather than just `(line, lineNum)`...

darcyparker commented on src/TsTypes.ts:
```diff
@@ -1,5 +1,9 @@
 import { camelCase, upperFirst } from 'lodash'
 
+const multiLineCommentStart =  '/** '
+const multiLineCommentIndent = '    '
```
BTW: If you like multi line comments like this:
```
/**
  *
  *
*/
```

its a simple change... Just define: `const multiLineCommentIndent = '  * '`

bcherny replied:
Bummer- I was trying to find a comment syntax that VSCode will use to inform its intellisense, but it doesn't look like comments for interfaces are supported.

donaldpipowitch replied:
I just tried to create a new issue to ask for a feature request to only use jsdoc comments for `descriptions`, before I saw this...

I don't know how you tried it, but VSCode works wonderfully if you use `/** */` in interfaces. It would be _really_ useful, if we could move descriptions to `/** */` comments.
![image](https://cloud.githubusercontent.com/assets/1152805/18661216/e2d70eca-7f14-11e6-9fae-4bd60d6fc938.png)

darcyparker replied:
I am using vim with https://github.com/Quramy/tsuquyomi and some other typescript plugins to get similar code completion, hints, jump to definition/references, etc..  Under the covers, the plugin uses a service from `tsc` and I guess VSCode uses this same service.

I found the hinting works when the interface is in the same file. But if it is imported, the hint points to the import. I have to jump to the definition to see the comment.  Perhaps this is the difference in @donaldpipowitch and @bcherny test?  Maybe one day `tsc` will be enhanced to provide hints to interfaces that are external to the current file?

@bcherny - what do you think? Should I retire the code that does the single line description?  I agree with @donaldpipowitch that it is really useful to have this info available to editors.

bcherny replied:
Ah, different use cases. When you're typing out `bar`, it should the description. But after you've entered `bar: someValue` and hover over `bar` again, the description is lost.

Anyway, @darcyparker I agree - if you want to tackle it in a separate commit in this PR, feel free to get rid of single line comments in favor of multiline jsdoc-style comments for intellisense. Otherwise, feel free to do it in a separate PR, or just file an issue.

darcyparker replied:
@bcherny - Done. I put it in same pull request, but two separate commits. I also rebased on top of latest `master` branch. See https://github.com/bcherny/json-schema-to-typescript/pull/26/commits/aef14ff462a106b76eadd0be8babc08e26fec14b

darcyparker commented on src/TsTypes.ts:
```diff
@@ -181,14 +185,26 @@ export namespace TsType {
+          //Multiple line description in comment above
           if (settings.propertyDescription && _.type.description && !_.type.id)
-            decl += ' // ' + _.type.description
+            decl = (_.type.description.search(newLineRegex) > 0) ?
```
Good catch. I updated to use `test()`.

darcyparker commented:
Please see https://github.com/bcherny/json-schema-to-typescript/pull/26/commits/0d5466d4f23fba17a7b08fb463111db2289646e1 I rebased ontop of the latest `master` branch and squashed because the rebase changed the commit ids.  @bcherny - your feedback about using `test()` is incorporated.

donaldpipowitch commented:
This is great. Thank you.
